OrbitQt/servicedeploymanager.h

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_SERVICE_DEPLOY_MANAGER_H_
+#define ORBIT_QT_SERVICE_DEPLOY_MANAGER_H_
+
+#include <QEventLoop>
+#include <QObject>
+#include <QPointer>
+#include <string>
+
+#include "OrbitSshQt/Session.h"
+#include "OrbitSshQt/SftpChannel.h"
+#include "OrbitSshQt/SftpOperation.h"
+#include "OrbitSshQt/Task.h"
+#include "OrbitSshQt/Tunnel.h"
+#include "deploymentconfigurations.h"
+#include "eventloop.h"
+
+namespace OrbitQt {
+
+class ServiceDeployManager : public QObject {
+  Q_OBJECT
+
+ public:
+  struct Ports {
+    uint16_t asio_port;
+    uint16_t grpc_port;
+  };
+
+  explicit ServiceDeployManager(
+      DeploymentConfiguration deployment_configuration,
+      OrbitSsh::Credentials creds, Ports remote_ports,
+      QObject* parent = nullptr);
+
+  outcome::result<Ports> Exec();
+  void Shutdown();
+  void Cancel();
+
+ signals:
+  void statusMessage(const QString&);
+  void socketErrorOccurred(std::error_code);
+
+ private:
+  EventLoop loop_;
+
+  DeploymentConfiguration deployment_configuration_;
+  OrbitSsh::Credentials credentials_;
+  Ports remote_ports_;
+  std::optional<OrbitSsh::Context> context_;
+  std::optional<OrbitSshQt::Session> session_;
+  std::optional<OrbitSshQt::Task> orbit_service_task_;
+  std::optional<OrbitSshQt::Tunnel> asio_tunnel_;
+  std::optional<OrbitSshQt::Tunnel> grpc_tunnel_;
+
+  outcome::result<void> ConnectToServer();
+  outcome::result<bool> CheckIfInstalled();
+  outcome::result<void> CopyOrbitServicePackage();
+  outcome::result<void> CopyOrbitServiceExecutable();
+  outcome::result<void> InstallOrbitServicePackage();
+  outcome::result<void> StartOrbitService();
+  outcome::result<void> StartOrbitServicePrivileged();
+  outcome::result<uint16_t> StartTunnel(
+      std::optional<OrbitSshQt::Tunnel>* tunnel, uint16_t port);
+
+  outcome::result<void> StartSftpChannel(OrbitSshQt::SftpChannel*);
+  outcome::result<void> CopyFileToRemote(
+      OrbitSshQt::SftpChannel&, std::string source, std::string dest,
+      OrbitSshQt::SftpOperation::FileMode dest_mode);
+  outcome::result<void> StopSftpChannel(OrbitSshQt::SftpChannel*);
+
+  void ShutdownSession();
+  void ShutdownOrbitService();
+  void ShutdownTunnel(std::optional<OrbitSshQt::Tunnel>*);
+
+  void handleSocketError(std::error_code);
+
+  template <typename Func>
+  [[nodiscard]] OrbitSshQt::ScopedConnection ConnectQuitHandler(
```

Comment:
are these nodiscards needed here?

Reply:
Let's put it like this: When the caller of this function discards the return value it's definitely not doing what he is trying to do and it's a bit subtle here as well.

But is it needed? Of course it would work without the nodiscard.